### PR TITLE
Fastsync fix

### PR DESCRIPTION
**Pull request by ebuchman**
- refactor some of the bash tests
- add test for fast sync
- fix fast sync

### CODE REVIEW

**Comment by ebuchman** [comment]
In future, we should use one call to ApplyBlock instead of the ExecBlock, Commit, Save sequence we have here. See https://github.com/tendermint/tendermint/pull/265

File: blockchain/reactor.go [hunk 1]
```diff
@@ -231,19 +231,22 @@ FOR_LOOP:
 					break SYNC_LOOP
 				} else {
 					bcR.pool.PopRequest()
+					// TODO: use ApplyBlock instead of Exec/Commit/SetAppHash/Save
```
Comment (jaekwon) [inline blockchain/reactor.go]: ApplyBlock?

Comment (ebuchman) [inline blockchain/reactor.go]: see handshake branch. it basically puts the ExecBlock, Commit, and setting of state root in one function

**Comment by jaekwon** [comment]
LGTM

**Pull request merged**